src/shared/openssl-util.c

```diff
@@ -1298,6 +1298,111 @@ int pkey_generate_volume_keys(
+        if (!info)
+                return log_openssl_errors("Failed to load OpenSSL store via '%s'", private_key_uri);
+
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *private_key = OSSL_STORE_INFO_get1_PKEY(info);
+        if (!private_key)
+                return log_openssl_errors("Failed to load private key via '%s'", private_key_uri);
+
+        *ret = TAKE_PTR(private_key);
+
+        return 0;
+#else
+        return -EOPNOTSUPP;
+#endif
+}
+
+DISABLE_WARNING_DEPRECATED_DECLARATIONS
+DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(ENGINE*, ENGINE_free, NULL);
+REENABLE_WARNING
+
+static int load_key_from_engine(const char *engine, const char *private_key_uri, EVP_PKEY **ret) {
+
+        assert(engine);
+        assert(private_key_uri);
+        assert(ret);
+
+        DISABLE_WARNING_DEPRECATED_DECLARATIONS
+        _cleanup_(ENGINE_freep) ENGINE *e = ENGINE_by_id(engine);
+        if (!e)
+                return log_openssl_errors("Failed to load signing engine '%s'", engine);
+
+        if (ENGINE_init(e) == 0)
+                return log_openssl_errors("Failed to initialize signing engine '%s'", engine);
+
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *private_key = ENGINE_load_private_key(
+                        e,
+                        private_key_uri,
+                        /* ui_method= */ NULL,
+                        /* callback_data= */ NULL);
+        if (!private_key)
+                return log_openssl_errors("Failed to load private key from '%s'", private_key_uri);
+        REENABLE_WARNING
+
+        *ret = TAKE_PTR(private_key);
+
+        return 0;
+}
+
+int openssl_load_key_from_token(const char *private_key_uri, EVP_PKEY **ret) {
+        _cleanup_free_ char *provider = NULL;
+        const char *colon, *e;
+        int r;
+
+        assert(private_key_uri);
+
+        colon = strchr(private_key_uri, ':');
+        if (!colon)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid URI '%s'", private_key_uri);
+
+        provider = strndup(private_key_uri, colon - private_key_uri);
+        if (!provider)
+                return log_oom_debug();
+
+        e = secure_getenv("SYSTEMD_OPENSSL_KEY_LOADER");
+        if (e) {
+                if (streq(e, "provider"))
+                        r = load_key_from_provider(provider, private_key_uri, ret);
+                else if (streq(e, "engine"))
+                        r = load_key_from_engine(provider, private_key_uri, ret);
+                else
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid value for SYSTEMD_OPENSSL_KEY_LOADER: %s", e);
+        } else {
+                r = load_key_from_provider(provider, private_key_uri, ret);
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to load key from provider '%s', falling back to engine", provider);
+                        r = load_key_from_engine(provider, private_key_uri, ret);
+                }
```

Comment:
Hmm, thinking more about it, I really dislike doing the fallback regardless of what error we get. Would it be OK for you if we by default used the providers and using the engine requires setting the environment variable? Passing through the environment variable to systemd-repart in mkosi is as simple as setting it in `Environment=`.

Reply:
The providers do not work at all, so basically it means that other than the command line option, the env var must always be set for this to have any meaning and usefulness, otherwise it's basically dead code. Even when they eventually start to work at some indeterminate point in the future, there's never going to be a clear error, just "i/o error". At that point I would question why add support for providers in the first place, and would just do engines only instead, given they actually work

Reply:
It's just some fallback for some external wonky API, why does it even matter?